### PR TITLE
Use helm directly

### PR DESCRIPTION
**Pull request by michaljurecko**
Part of: https://keboola.atlassian.net/browse/BA-18

Changes:
- `etcd` is deployed by Helm.
- Deploy (old cluster state -> new cluster state) is tested in the CI in the `MiniKube`.
- Updated `etcd` base image `3.5.4` -> `3.5.5`.

### CODE REVIEW

File: provisioning/templates-api/kubernetes/templates/etcd-values.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # See default values: https://github.com/bitnami/charts/blob/master/bitnami/etcd/values.yaml
 image:
-  tag: 3.5.4-debian-11-r27
+  tag: 3.5.5-debian-11-r16
```
Comment (michaljurecko) [inline provisioning/templates-api/kubernetes/templates/etcd-values.yaml:3]: Ked to nasadzujeme znova, tak som rovno updatoval image.
(Usetri mi to jeden deploy, pokryvaju to testy.)

File: provisioning/templates-api/kubernetes/templates/etcd-values.yaml [hunk 2]
```diff
@@ -28,7 +28,7 @@ auth:
     create: true
     allowNoneAuthentication: false
   token:
-    type: simple
+    type: jwt
```
Comment (michaljurecko) [inline provisioning/templates-api/kubernetes/templates/etcd-values.yaml:31]: Bez Helm-u sme nevedeli nasetupovat defaultne `jwt` tokeny (ktore si etcd generuje interne).,
Teraz to uz ide, Helm vygeneruje private key.

Je to rychlejsi a bezpecnejsi sposob, ako `simple`.
Je to aj default hodnota: 
https://github.com/bitnami/charts/blob/638a48e4d3ec7b5d160f4b525ec40218512c464b/bitnami/etcd/values.yaml#L119-L131

image


------

Novy secret:

image

File: .github/workflows/templates-api.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+name: 'Provisioning: Templates API'
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply .github/workflows/templates-api.yml:1]: Tu je vysledok testu: https://github.com/keboola/keboola-as-code/actions/runs/3362829865/jobs/5575086492

image
image
------------
image
image
----------
image

**Comment by michaljurecko** [comment]
Vyskusal som aj manualne, ze `etcd locks` po tychto zmenach funguju spravne:
image

**Comment by michaljurecko** [comment]
(Okomentoval som to po commitoch).

**Review by JakubMatejka — APPROVED** [review]
Za mě ok, budeš to dávat na review SRE?

**Comment by michaljurecko** [comment]
@JakubMatejka Neplanoval som to davat na review SRE.
Nevidim tam nejaku zasadnu zmenu v tom co je nasadene, ale iba v tom ako to nasadzujeme.
Ale teda, asi tym chces povedat, ze by som mal, tak dam.

@Halama , @ondrejhlavacek, mali by ste niekto z SRE chvilku cas na review?

Ak to nechcete pozerat cele, tak tu je diff, stara verzia (`envsubstr`) vs nova verzia (`helm`).
https://www.diffchecker.com/wM53h8ks

Ziskane pomocou:
```
helm template templates-api-etcd bitnami/etcd -f ./kubernetes/deploy/etcd-values.yaml --namespace templates-api --set auth.rbac.rootPassword=$ETCD_ROOT_PASSWORD > out.yaml
```

Kedze v kode mame:
```
helm upgrade --install templates-api-etcd bitnami/etcd -f ./kubernetes/deploy/etcd-values.yaml --namespace templates-api --set auth.rbac.rootPassword=$ETCD_ROOT_PASSWORD
```

Vyhody:
- Mozeme pouzit rychlejsie a bezpecnejsie `jwt` tokeny - Helm vygeneruje private key (prvy krat).
- Jednoduchsie sa nam to konfiguruje cez `helm`, bude z toho vychadzat `etcd` pre buffer API s nejakymi dalsimi nastaveniami (v dalsich PRs).

**Comment by Halama** [comment]
Hodil bych na to rychle oko tak behem hodky snad.

**Comment by JakubMatejka** [comment]
@michaljurecko já myslel aspoň formálně, když se to týká infra. 🙂

File: provisioning/templates-api/kubernetes/templates/etcd-values.yaml (hunk 2 above)
Comment (Halama) [reply provisioning/templates-api/kubernetes/templates/etcd-values.yaml:31]: Rozumim tomu spravne tak ze pro prvotni autentizaci se pouzije jmeno/heslo https://github.com/keboola/keboola-as-code/blob/75923882f959e6f06353e4c0153be6be1507bd6a/provisioning/templates-api/kubernetes/templates/templates-api.yaml#L51 ale pak uz se autentizuje [jwt tokenem](https://etcd.io/docs/v3.5/learning/design-auth-v3/#authentication)?

File: provisioning/templates-api/common.sh [hunk 4]
```diff
@@ -16,9 +16,22 @@ fi
...
+
+# REMOVE ME: delete old manually deployed resources
+kubectl delete pod,statefulset,configmap,networkpolicy,secret,service,persistentVolumeClaim,persistentVolume --namespace templates-api -l app=templates-api-etcd --ignore-not-found
+
+# Get etcd root password, if it is already present
+export ETCD_ROOT_PASSWORD=$(kubectl get secret --namespace "templates-api" templates-api-etcd -o jsonpath="{.data.etcd-root-password}" 2>/dev/null | base64 -d)
+
+# Deploy etcd cluster
+helm repo add --force-update bitnami https://charts.bitnami.com/bitnami
+helm upgrade --install templates-api-etcd bitnami/etcd -f ./kubernetes/deploy/etcd-values.yaml --namespace templates-api --set auth.rbac.rootPassword=$ETCD_ROOT_PASSWORD
```
[the remark this answers is not in the document]
Comment (Halama) [reply provisioning/templates-api/common.sh:33]: Neměla by tu radši být nějak zafixovaná verze chartu?

**Review by Halama — APPROVED** [review]
Rychle jsem to prošel, myslím že to vypadá ok.

**Comment by michaljurecko** [comment]
@Halama diky za review:
- zamkol som verziu https://github.com/keboola/keboola-as-code/pull/936/commits/82b62749c7259701723e8dcc238da4a7e873129d
- ano, tie `jwt` chapes spravne + pouzivaju sa aj na komunikaciu medzi `etcd` nodes
  - nas klient ziska kazdych `10m` pomocou hesla novy token, ktory uz potom pouziva pri dalsich operaciach

-----
Z etcd docs:

**Two types of tokens: simple and JWT**

There are two kinds of token types: simple and JWT. **The simple token isn't designed for production use cases.** Its tokens aren't cryptographically signed and servers must statefully track token-user correspondence; it is meant for development testing. JWT tokens should be used for production deployments since it is cryptographically signed and verified. From the implementation perspective, JWT is stateless. Its token can include metadata including username and revision, so servers don't need to remember correspondence between tokens and the metadata.

**Pull request merged**